client/app/pages/dashboards/dashboard.js

```diff
@@ -3,9 +3,58 @@ import template from './dashboard.html';
 import shareDashboardTemplate from './share-dashboard.html';
 
 function DashboardCtrl(
-  $rootScope, $routeParams, $location, $timeout, $q, $uibModal,
+  $rootScope, $scope, $routeParams, $location, $timeout, $q, $uibModal,
   Title, AlertDialog, Dashboard, currentUser, clientConfig, Events,
+  dashboardGridOptions, toastr,
 ) {
+  this.saveInProgress = false;
+  const saveDashboard = _.debounce(() => {
+    if (!this.dashboard.canEdit()) {
+      return;
+    }
+
+    this.saveInProgress = true;
+    const showMessages = this.layoutEditing;
+    // Temporarily disable grid editing (but allow user to use UI controls)
+    this.dashboardGridOptions.draggable.enabled = false;
+    this.dashboardGridOptions.resizable.enabled = false;
+    $q.all(_.map(this.dashboard.widgets, widget => widget.$save()))
+      .then(() => {
+        if (showMessages) {
+          toastr.success('Dashboard layout saved.');
+        }
+      })
+      .catch(() => {
+        if (showMessages) {
+          toastr.error('Cannot save dashboard layout.');
+        }
+      })
+      .finally(() => {
+        this.saveInProgress = false;
+        // If user didn't disable editing mode while saving - restore grid
+        this.dashboardGridOptions.draggable.enabled = this.layoutEditing;
+        this.dashboardGridOptions.resizable.enabled = this.layoutEditing;
+      });
+  }, 2000);
+
+  this.layoutEditing = false;
+  this.dashboardGridOptions = _.extend({}, dashboardGridOptions, {
+    resizable: {
+      enabled: false,
+      handles: ['n', 'e', 's', 'w', 'ne', 'se', 'sw', 'nw'],
+      stop: saveDashboard,
+    },
+    draggable: {
+      enabled: false,
+      stop: saveDashboard,
+    },
+  });
+
+  $scope.$watch(
```

Comment:
We try to move away from using `$watch` whenever possible. In this case, I believe there are two cases when the widgets list might change: when a widget is added or when it's removed. We can have callbacks in the code for both cases instead of constantly checking the state of the array.

Another question is why do we need to save the dashboard in such case? To update the locations of the widgets?

Reply:
Yes, because when we add/remove widgets, other widgets' positions may change, so we need to save it all. I know about cons of using `$watch`, therefore I'm just watching `length` property, not entire list of objects, - it's fast and will not lead to performance issues. My opinion is - if we need this in the single place, and is clear and fast enough - it can be implemented in (such) a simple way.

Reply:
There is the performance concern, which you address nicely with watching only the length. But there are other concerns of maintainability and making the code clear -- future reader might not understand why we need this.

But there is a worse issue here: because of the watch, `saveDashboard` is called every time you load the dashboard.

We already have everything "provisioned" to run code after adding a widget:

https://github.com/getredash/redash/blob/900d5588578eae027dd7a6caf5f91383d8893da2/client/app/pages/dashboards/dashboard.js#L181

And after delete:

https://github.com/getredash/redash/blob/900d5588578eae027dd7a6caf5f91383d8893da2/client/app/pages/dashboards/dashboard.html#L65

Maybe add new controller function for post widgets change tasks and run there `extractGlobalParams` and `saveDashboard`.

Reply:
Btw, another reason why I try to stay away from things like `$watch`, is that it makes the code more portable to other frameworks. 😉

Reply:
You're right. Will think how to improve this code.